REFACTORING_RECOMMENDATIONS.md

```diff
@@ -0,0 +1,1205 @@
+    /**
+     * Encrypts payload with automatic nonce generation
+     */
+    static encode(
+        encryptionKey: Bytes,
+        payload: Bytes,
+        messageCounter: number,
+        nodeId?: NodeId
+    ): Bytes {
+        const nonce = this.generateNonce(messageCounter, nodeId);
+        return Crypto.encrypt(encryptionKey, payload, nonce);
+    }
+
+    /**
+     * Decrypts encrypted data, extracting nonce from packet
+     */
+    static decode(
+        decryptionKey: Bytes,
+        encrypted: Bytes,
+        messageCounter: number,
+        nodeId?: NodeId
+    ): Bytes {
+        const nonce = this.generateNonce(messageCounter, nodeId);
+        return Crypto.decrypt(decryptionKey, encrypted, nonce);
+    }
+
+    /**
+     * Generates nonce from message counter and optional node ID
+     */
+    private static generateNonce(messageCounter: number, nodeId?: NodeId): Bytes {
+        const nonce = new Uint8Array(13);
+        const view = new DataView(nonce.buffer);
+
+        // Write message counter (4 bytes, little-endian)
+        view.setUint32(0, messageCounter, true);
+
+        // Write node ID if provided (8 bytes, little-endian)
+        if (nodeId !== undefined) {
+            view.setBigUint64(5, nodeId, true);
+        }
+
+        return nonce;
+    }
+}
+```
+
+**Refactor Sessions**:
+
+```typescript
+// Before: /packages/protocol/src/session/NodeSession.ts
+encode(payload: Bytes): Bytes {
+    const nonce = this.#generateNonce();
+    return Crypto.encrypt(this.encryptionKey, payload, nonce, ...);
+}
+
+#generateNonce(): Bytes {
+    // ... 10+ lines of nonce generation
+}
+
+// After:
+import { SessionCrypto } from "./SessionCrypto.js";
+
+encode(payload: Bytes): Bytes {
+    return SessionCrypto.encode(
+        this.encryptionKey,
+        payload,
+        this.messageCounter,
+        this.nodeId
+    );
+}
+```
+
+**Testing**:
+- Unit tests for `SessionCrypto`
+- Test nonce generation consistency
+- Test encryption/decryption round-trip
+- Verify compatibility with existing sessions
+- Performance tests
+
+**Files to Refactor**:
+1. `session/NodeSession.ts:236-269`
+2. `session/GroupSession.ts:125-216`
+
+---
+
+### 2.5 Create BDX Schema Factory
+
+**Files Affected**: 5+ schema classes
+**Lines Saved**: ~75
+**Effort**: 6-8 hours
+
+**Implementation**:
+
+```typescript
+// File: /packages/protocol/src/bdx/schema/SimpleSchemaFactory.ts
+import { Schema } from "#types";
+import { DataReader, DataWriter, Endian } from "#general";
+import { Bytes } from "#general";
+
+type FieldType = 'UInt8' | 'UInt16' | 'UInt32' | 'UInt64' | 'ByteArray';
+
+interface FieldDefinition<T> {
+    key: keyof T;
+    type: FieldType;
+    /** Optional length for ByteArray type */
+    length?: number;
+}
+
+/**
+ * Creates a simple schema for BDX messages with automatic encoding/decoding.
+ * Eliminates boilerplate for schemas with primitive field types.
+ */
+export function createSimpleSchema<T extends Record<string, any>>(
```

Comment:
Interesting but this didn't seem BDX specific. 

Also prefer names like `SimpleSchema` over `createSimpleScema` for factories